pkg/topology/kademlia/internal/metrics/metrics.go

```diff
@@ -135,77 +109,56 @@ func (ss *Snapshot) HasAtMaxOneConnectionAttempt() bool {
 	return ss.LastSeenTimestamp == 0 && ss.SessionConnectionRetry <= 1
 }
 
+// persistentCounters is a helper struct used for persisting selected counters.
+type persistentCounters struct {
+	PeerAddress       *swarm.Address `json:"peerAddress"`
+	LastSeenTimestamp int64          `json:"lastSeenTimestamp"`
+	ConnTotalDuration time.Duration  `json:"connTotalDuration"`
+}
+
 // Counters represents a collection of peer metrics
 // mainly collected for statistics and debugging.
 type Counters struct {
 	sync.Mutex
 
 	// Bookkeeping.
-	peer     *swarm.Address
-	loggedIn bool
-
-	// Watches in-memory counters which has to be persisted.
-	// 	3 - dirty, need to be persisted
-	//	2 - snapshot of counters in progress
-	//	1 - batched for persistent write
-	//	0 - persisted
-	dirty atomic.Int32
+	isLoggedIn  bool
+	peerAddress *swarm.Address
 
-	// In-memory counters.
+	// Counters.
 	lastSeenTimestamp    int64
 	connTotalDuration    time.Duration
 	sessionConnRetry     uint64
 	sessionConnDuration  time.Duration
 	sessionConnDirection PeerConnectionDirection
-
-	// Persistent counters.
-	persistentLastSeenTimestamp atomic.Value
-	persistentConnTotalDuration atomic.Value
 }
 
-// flush writes the current state of in memory counters into the given db.
-func (cs *Counters) flush(db *shed.DB, batch *leveldb.Batch) error {
-	if cs.dirty.Load() < 3 {
-		return nil
+// UnmarshalJSON unmarshal just the persistent counters.
+func (cs *Counters) UnmarshalJSON(b []byte) (err error) {
+	var val persistentCounters
+	if err := json.Unmarshal(b, &val); err != nil {
+		return err
 	}
-	cs.dirty.CAS(3, 2)
-
-	cs.Lock()
-	var (
-		key                             = cs.peer.String()
-		lastSeenTimestampSnapshot       = cs.lastSeenTimestamp
-		connectionTotalDurationSnapshot = cs.connTotalDuration
-	)
-	cs.Unlock()
-
-	ls, ok := cs.persistentLastSeenTimestamp.Load().(*shed.Uint64Field)
-	if !ok {
-		mk := newPeerKey(peerLastSeenTimestamp, key)
-		field, err := db.NewUint64Field(mk.String())
-		if err != nil {
-			return fmt.Errorf("field initialization for %q failed: %w", mk, err)
-		}
-		ls = &field
-		cs.persistentLastSeenTimestamp.Store(ls)
+	if cs != nil {
```

Comment:
is this meant to protect against a panic? if so, it also swallows an error and does not indicate that a programmer error has occurred. i would suggest to remove the if statement and directly access the fields. is there any case in which this is `nil`?